src/services/intelligentTests.js

```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env node
+
+const shellJS = require('shelljs');
+const fs = require('fs');
+const logger = require('../utils/logger');
+
+const getTestsFromTargetOrg = (apexClassesList) => new Promise((resolve, reject) => {
+    const codeCoverageQuery = `SELECT ApexTestClass.Name
+                               FROM ApexCodeCoverage
+                               WHERE ApexClassorTrigger.Name IN (${apexClassesList.map(name => `'${name}'`).join(', ')})`;
+    const target = 'AVG_SFQA';
+    const queryRes = shellJS.exec(`sfdx force:data:soql:query -t -q "${codeCoverageQuery}" -u ${target} -r json`, {silent: true});
+
+    const apexClassNameWithId = JSON.parse(queryRes.stdout);
+    resolve(new Set(apexClassNameWithId.result.records.map(payload => payload.ApexTestClass.Name)));
+});
+
+const getApexClassesFromArticlePath = (articlePath, projectType) => new Promise((resolve, reject) => {
+    let apexClassesQuery;
+    if (projectType === 'mdapi') {
+        apexClassesQuery = `${articlePath}/classes/`;
+    } else if (projectType === 'source') {
+        // TODO: need to account for non-default package
+        apexClassesQuery = `${articlePath}/force-app/main/default/classes/`;
+    } else {
+        logger.error(`Provided projectType: ${projectType} is not recognized`);
+        reject();
+    }
+    //clean extra /'s
+    apexClassesQuery.replaceAll('//', '/');
+
+    const listOfClasses = fs.readdirSync(apexClassesQuery);
+
+    const categorizedClasses = {
+        Classes: [],
+        Tests: []
+    };
+
+    listOfClasses.forEach(filePath => {
```

Comment:
@henrymzhao You are trying to categorize test classes and normal classes and I believe eventually you are utilizing the normal classes in your Apex Code Coverage query. Even if you do not categorize and provide the whole dump of classes in the commit or git diff, it may still be OK as Test classes in the query will not provide any list of test classes. This whole logic of grep etc. may not be required then

Reply:
Yes for that I was thinking about future enhancements. There are cases where (and it's bad practice) people create logic classes then put test methods within those logic classes themselves. Eventually I'm thinking of a way to identify those with grep and create a more accurate set of all tests.

But you're right, I could just include them all in the query since non-test classes won't break the query anyways, that's a good call.